File: Trees/Sum Root to Leaf Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int sumNumbers(TreeNode* root) {
        return recSum(root, 0);
    }

    int recSum(TreeNode* root, int prevSum) {
        int currSum = prevSum*10 + root->val;
        
        if (root->left == nullptr && root->right == nullptr) {
            return prevSum*10 + root->val;
        } else if (root->right == nullptr) {
            return recSum(root->left, prevSum*10 + root->val);
        } else if (root->left == nullptr) {
            return recSum(root->right, prevSum*10 + root->val);
        } else {
            return recSum(root->left, prevSum*10 + root->val) + recSum(root->right, prevSum*10 + root->val);
        }
    }
};

File: Trees/Sum Root to Leaf Numbers.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn sum_numbers(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        Self::helper(&root.unwrap(), 0)
    }

    fn helper(root: &Rc<RefCell<TreeNode>>, prev_sum: i32) -> i32 {
        let curr_sum = prev_sum * 10 + root.borrow().val;

        match (&root.borrow().left, &root.borrow().right) {
            (Some(left), Some(right)) => {
                Self::helper(left, curr_sum) + Self::helper(right, curr_sum)
            },
            (Some(left), _) => {
                Self::helper(left, curr_sum)
            },
            (_, Some(right)) => {
                Self::helper(right, curr_sum)
            },
            (None, None) => {
                curr_sum
            }
        }
    }
}